src/java/org/apache/cassandra/utils/btree/BTree.java

```diff
@@ -1448,4 +1920,2355 @@ private static boolean isStopSentinel(long v)
+            }
+            else
+            {
+                if (savedBuffer == null)
+                {
+                    savedBuffer = new Object[2 * (1 + MAX_KEYS)];
+                    savedSizes = new int[1 + MAX_KEYS];
+                }
+
+                System.arraycopy(  pred,    0,        savedBuffer, 0,        predKeys);
+                System.arraycopy(  pred,    predKeys, savedBuffer, MAX_KEYS, predKeys + 1);
+                copySizeMapToSizes(sizeMap, 0,        savedSizes,  0,        predKeys + 1);
+                if (newKeys == MAX_KEYS + 1)
+                {
+                    savedNextKey = predNextKey;
+                }
+                else
+                {
+                    int removeKeys = (1 + MAX_KEYS - newKeys);
+                    int remainingKeys = count - removeKeys;
+
+                    savedBuffer [predKeys] = predNextKey;
+                    System.arraycopy(buffer, 0,                     savedBuffer, newKeys,            MAX_KEYS     - newKeys);
+                    savedNextKey =            buffer[MAX_KEYS     - newKeys];
+                    System.arraycopy(sizes,  0,                     savedSizes,  newKeys,            MAX_KEYS + 1 - newKeys);
+                    System.arraycopy(buffer, MAX_KEYS,              savedBuffer, MAX_KEYS + newKeys, MAX_KEYS + 1 - newKeys);
+                    System.arraycopy(buffer, removeKeys,            buffer,      0,                  remainingKeys);
+                    System.arraycopy(buffer, MAX_KEYS + removeKeys, buffer,      MAX_KEYS,           remainingKeys + 1);
+                    System.arraycopy(sizes,  removeKeys,            sizes,       0,                  remainingKeys + 1);
+                    count = remainingKeys;
+                }
+            }
+        }
+
+        boolean producesOnlyDense()
+        {
+            return sizes == null;
+        }
+
+        /**
+         * Construct a new branch from the contents of {@code branchBuffers[branchIndex]}, unless the contents have
+         * not changed from {@code unode}, in which case return {@code unode} to avoid allocating unnecessary objects.
+         */
+        Object[] drainAndPropagate(Object[] unode, int usz, BranchBuilder propagateTo)
+        {
+            int sizeOfBranch;
+            Object[] branch;
+            if (mustRedistribute())
+            {
+                branch = redistributeOverflowAndDrain();
+                sizeOfBranch = sizeOfBranch(branch);
+            }
+            else if (!hasOverflow() && usz == count
+                     && areIdentical(buffer, 0,        unode, 0,   usz)
+                     && areIdentical(buffer, MAX_KEYS, unode, usz, usz + 1))
+            {
+                branch = unode;
+                sizeOfBranch = sizeOfBranch(branch);
+            }
+            else
+            {
+                if (hasOverflow())
+                    propagateOverflow();
+
+                branch = new Object[2 * (count + 1)];
```

Comment:
It can be, I think. If this leaf is the root. The `mustRedistribute()` path above captures all other cases, no? 

When we are building a new tree, and we are not the root and so have overflowed at least once, we always maintain one full node at each level we are constructing plus some additional non-empty buffer.

In the case where we are inserting into an existing tree, we always have at least `MIN_KEYS` input on the side we are inserting into, so we must have at least MIN_KEYS+1 in the output.

In the case where we are transforming a tree and may remove items, we scavenge elements to ensure we always have MIN_KEYS.

Reply:
This is correct, let's add a comment to that effect.